Merge/foundation release/1.12.2+spiral

meowsbits opened the pull request:
Implements the pending Spiral ([ECIP-1109](https://ecips.ethereumclassic.org/ECIPs/ecip-1109))  network upgrade for ETC.

meowsbits committed:
wip

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params/types/coregeth,params/types/ctypes,params/types/genesisT,params/types/goethereum: install Shanghai=Spiral block number based iface methods

ETC plans to install most EIPs associated with
ETH's Shanghai fork, but ETC doesn't use PoS so activation
by block time, so this adds supporting block-based activation
methods for those features.
Still TODO is to actually implement the associated conditionals
where the Getters are used in the code to actually turn
the new features on.

Date: 2023-08-18 09:11:04-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params: (typo) remove spurious and meaningless line

Date: 2023-08-18 19:28:32-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params/types/coregeth: fix feature-block type: uint64/big.Int

Blocks are big ints.
Times are uint64s.

I am 73% certain this covers all cases.

Date: 2023-08-18 19:30:17-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core,core/vm/runtime,params,tests: GetEIP3651Transition

Date: 2023-08-18 19:50:22-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core,eth/catalyst: SetEIP3651Transition

Date: 2023-08-19 07:24:31-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
consensus/clique,core/vm: GetEIP3855Transition

Date: 2023-08-19 07:29:40-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth/catalyst: SetEIP3855Transition

Date: 2023-08-19 07:43:32-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
cmd/evm/internal/t8ntool,consensus/ethash,core,core/txpool,core/vm,light: GetEIP3860Transition

Date: 2023-08-19 07:53:58-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth/catalyst: SetEIP3860Transition

Date: 2023-08-19 07:55:11-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
cmd/evm/internal/t8ntool,consensus/beacon,core,eth/catalyst: GetEIP4895Transition

Date: 2023-08-19 08:22:35-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params: (typo) remove spurious character '$'

Date: 2023-08-19 08:23:35-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth/catalyst: SetEIP4895Transition

Date: 2023-08-19 08:27:25-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth/catalyst: SetEIP6049Transition

Date: 2023-08-19 08:30:24-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core/vm,params: GetEIP4399Transition (noop for ETC)

Date: 2023-08-19 08:54:33-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params: install Spiral placeholder activation vals

Date: 2023-08-19 09:13:58-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params/types/goethereum: fix incompatible check; permit passthrough block-based Shanghai config

Failing was TestCheckCompatible, which
converts Geth::CoreGeth configuration types;
CoreGeth's EIP3651 block transition was fatally
inapplicable to Geth's config.

So this patch sidesteps the issue by permitting
the conversion to quietly fail; expecting that users/developers
will be sane enough to configure upgrades with one pattern or the other
but not both.

Date: 2023-08-19 09:17:33-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
tests: add ETC_Spiral test fork config

Date: 2023-08-25 08:55:58-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Merge branch 'master' into spiral

meowsbits committed:
params/types/coregeth,params/types/goethereum: problem: TestEquivalent_Features fails

Mismatched EIP3651 activations caused an
incompatibility between core-geth and go-ethereum
config types.

This is mostly just a problem of/with the tests
because of the disparate configuration TYPES
now (block vs. time).

This patches fixes the issue by implementing
a logical short-circut equivocating zero values
for time and block fields in core-geth
and go-ethereum configuration configurator
interfaces.

Probably these logical steps will only be followed
by tests or testnets or networks using time OR
block values.

Date: 2023-08-25 14:51:50-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
params/types/genesisT: fix indentation

ziogaschr committed:
tests: set new head for tests/testdata

ziogaschr committed:
tests: Set new submodule head for tests/testdata-etc

meowsbits committed:
Merge branch 'spiral' into merge/foundation-release/1.12.2+spiral

Conflicts:
      core/genesis.go
      core/txpool/txpool.go
      core/vm/jump_table.go
      params/types/genesisT/genesis.go
      tests/init.go

meowsbits committed:
beacon/engine: include error in wrapped error Error method

By Go conventions the Error method should
return a descriptive string.

When I was debugging the simulated beacon
test, the reported error was hardly usedful,
describing only that an error existed, but lacking
any contextual detail.

Now the engine api will return a more
detailed error and that will hopefully
help debuggers in the future.

Date: 2023-08-29 09:01:11-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth/catalyst: fix EIP4895 conditional logic

This logic was just wrong because of
misplaced negation.

Date: 2023-08-29 09:10:51-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Revert "beacon/engine: include error in wrapped error Error method"

This reverts commit ae0efa109d37e8374999a304bb03dd3ece6bde2a.

Reason: it caused a test failure TestGetBlockBodiesByRangeInvalidParams
where that test compares the have/want errors as strings
and I can't be bothered to adjust the test.

It would be better, probably, to use errors.Is with wrapped
errors.

meowsbits committed:
Merge branch 'merge/foundation-release/1.12.2' into merge/foundation-release/1.12.2+spiral

ziogaschr commented on core/genesis.go:
```diff
@@ -436,7 +438,7 @@ func GenesisToBlock(g *genesisT.Genesis, db ethdb.Database) *types.Block {
 	var withdrawals []*types.Withdrawal
 	if conf := g.Config; conf != nil {
-		isShangai := conf.IsEnabledByTime(g.Config.GetEIP4895TransitionTime, &g.Timestamp)
+		isShangai := conf.IsEnabledByTime(g.Config.GetEIP4895TransitionTime, &g.Timestamp) || g.Config.IsEnabled(g.Config.GetEIP4895Transition, new(big.Int).SetUint64(g.Number))
```
Btw, I found out recently, that on the last release merge for some reason I used vars named as `isShangai` instead of `isEip4895`. I am thinking to revisit them and change them for consistency. What do you think guys?

meowsbits replied:
I like `isEIP4895` better, and even upstream is now following that pattern more and more. Comments about EIP::Forkname are always useful though )

ziogaschr replied:
I will handle those in a different PR. Resolving this conversation now.

ziogaschr commented on core/vm/jump_table.go:
```diff
@@ -193,7 +193,7 @@ func instructionSetForConfig(config ctypes.ChainConfigurator, isPostMerge bool,
 		enable3198(instructionSet) // BASEFEE opcode https://eips.ethereum.org/EIPS/eip-3198
 	}
-	if isPostMerge {
+	if isPostMerge || config.IsEnabled(config.GetEIP4399Transition, bn) { // EIP4399: Supplant DIFFICULTY opcode with PREVRANDAO (ETH @ PoS)
```
Nice catch

ziogaschr commented on eth/catalyst/api.go:
```diff
@@ -169,7 +170,8 @@ func (api *ConsensusAPI) ForkchoiceUpdatedV1(update engine.ForkchoiceStateV1, pa
 		}
-		if api.eth.BlockChain().Config().IsEnabledByTime(api.eth.BlockChain().Config().GetEIP4895TransitionTime, &payloadAttributes.Timestamp) {
+		if api.eth.BlockChain().Config().IsEnabledByTime(api.eth.BlockChain().Config().GetEIP4895TransitionTime, &payloadAttributes.Timestamp) || // PTAL(meowsbits) Is it really safe/sensible to call the backing Blockchain here?
+			api.eth.BlockChain().Config().IsEnabled(api.eth.BlockChain().Config().GetEIP4895Transition, api.eth.BlockChain().CurrentHeader().Number) {
```
Note to me, to revisit it

ziogaschr replied:
Fixed by 87843da

ziogaschr commented on params/types/coregeth/chain_config_configurator.go:
```diff
@@ -516,51 +516,147 @@ func (c *CoreGethChainConfig) SetEIP4399Transition(n *uint64) error {
+
+// Shanghai by block
+// EIP3651: Warm COINBASE
+func (c *CoreGethChainConfig) GetEIP3651Transition() *uint64 {
```
I like how the setters and getters of EIPs that can be activated with both time and block number check to each other, although I am a bit worried to what happens if both are being set by mistake and if the user will notice it.

Do you think that it is better if we take precedence of one over the other? For example, first check time then block number no matter which getter is being used.

Currently the siblings is only checked if enabled on genesis block, though some developers might mistakenly pass 0 values to some EIPs when they init chains using custom genesis.

meowsbits replied:
> Do you think that it is better if we take precedence of one over the other? For example, first check time then block number no matter which getter is being used.

The problem is that each getter can only represent the time OR block value; a single getter can't reconcile between the two kinds of activation values (except for the special `0` case, where `time:0 -> block:0`; even `block:0 -> time:0` is not logically sound, since genesis blocks can, and do, have non-`0` timestamps).

And I think the problem persists, were precedence to be established, with developers mistakenly passing `nil` values (true zero-values), the same with `0` values. 

But I'm not sure I've perfectly understood your proposal, nor that I've even provided a watertight reason -- maybe you can show a pseudocode or snippet of what you're considering?

ziogaschr replied:
@meowsbits I removed this logic as it's not needed. There was a copy/paste typo which caused all this. Here is the fix [26265c7](https://github.com/etclabscore/core-geth/pull/561/commits/26265c73c011f2e678f6aff4fc99db114cc5d4bf).

meowsbits committed:
beacon/engine,eth/catalyst: catalyst api payload attributes get block number

In order to check eip4895 features for ETC-style
block number-based activations, the attributes get
a block number field which is allowed to be nil.

As usual for core-geth, activation numbers are OR'd
with activation times.

Date: 2023-09-01 10:18:47-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core: prefer EIP descriptors over fork names

... but reference forks by name to
keep the important context.

Date: 2023-09-04 05:22:48-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
params/types/coregeth,params/types/goethereum: Revert "params/types/coregeth,params/types/goethereum: problem: TestEquivalent_Features fails"

This reverts commit a5e0c62fe940d93578105b8c8cd81688cd13f174.

ziogaschr committed:
params/types/coregeth: fix typo on core-geth chain configurator EIPs for Shanghai

ziogaschr reviewed (approved):
LGTM. Nice work man

meowsbits committed:
params: unset Spiral/Shanghai activation blocks for Classic, Mordor

The ECIP does not specify valid activation
blocks yet, so we should not code them.
https://github.com/ethereumclassic/ECIPs/blob/64423c73b7416c2f8e40d4c66bf40b715c458b8c/_specs/ecip-1109.md

Date: 2023-09-04 10:34:38-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
core/forkid: unset Spiral activation blocks tests for Classic
